### PR TITLE
Fixed #19580 -- Made nullable relations work consistently

### PR DESCRIPTION
**Pull request by akuzminov**
https://code.djangoproject.com/ticket/19580

### CODE REVIEW

**Comment by MarkusH** [comment]
buildbot, ok to test

File: tests/m2m_through_regress/tests.py [hunk 1]
```diff
@@ -251,6 +252,19 @@ def test_remove_reverse(self):
         self.assertQuerysetEqual(
             self.driver.car_set.all(), [])
 
+    def test_null_query(self):
+        cd = CarDriver(driver=self.driver)
+        c = Car()
+        with self.assertRaises(ValueError):
```
Comment (timgraham) [inline tests/m2m_through_regress/tests.py]: use assertRaisesMessage

File: tests/m2m_through_regress/tests.py [hunk 2]
```diff
@@ -251,6 +252,19 @@ def test_remove_reverse(self):
...
+            c.drivers.all()
+        c.save()
+        self.assertTrue(isinstance(c.drivers.all(), EmptyQuerySet))
```
Comment (timgraham) [inline tests/m2m_through_regress/tests.py]: use assertIsInstance

**Comment by timgraham** [comment]
Looks like some more tests are needed. There should be a test that fails if any of the lines in `related_descriptors.py` are reverted.

File: django/db/models/fields/related_descriptors.py [hunk 3]
```diff
@@ -875,8 +875,16 @@ def get_queryset(self):
                 return self.instance._prefetched_objects_cache[self.prefetch_cache_name]
             except (AttributeError, KeyError):
                 queryset = super(ManyRelatedManager, self).get_queryset()
+                if self.instance.pk is None or None in self.related_val:
+                    return queryset.none()
                 return self._apply_rel_filters(queryset)
 
+        def _check_has_fk_val(self):
+            if self.instance.pk is None or None in self.related_val:
+                raise ValueError('"%r" needs to have a value for field "%s" before '
```
Comment (timgraham) [inline django/db/models/fields/related_descriptors.py]: Use hanging indent:
```
raise ValueError(
    '"%r" needs to have a value for field "%s" before '
    ...
)
```

File: tests/m2m_through_regress/tests.py [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 from django.contrib.auth.models import User
 from django.core import management
+from django.db.models.query import EmptyQuerySet
```
Comment (timgraham) [inline tests/m2m_through_regress/tests.py]: from django.core.exceptions ...

**Comment by timgraham** [comment]
Hi, do you have time to work on this soon or should we release the ticket and let someone else continue it? Thanks.

**Comment by akuzminov** [comment]
Hey, I've just got back from overseas. Will have time to finish it this week

**Comment by akuzminov** [comment]
@timgraham I have updated the pull request

File: tests/m2m_through_regress/tests.py [hunk 5]
```diff
@@ -251,6 +260,43 @@ def test_remove_reverse(self):
         self.assertQuerysetEqual(
             self.driver.car_set.all(), [])
 
+    def test_remove_empty(self):
+        c = Car.objects.create(make=None)
+        c.save()
```
Comment (olibook) [inline tests/m2m_through_regress/tests.py]: save is not needed : objects.create returns an already saved instance

File: tests/m2m_through_regress/tests.py [hunk 6]
```diff
@@ -251,6 +260,43 @@ def test_remove_reverse(self):
...
+        with self.assertRaisesMessage(
+            ValueError,
+            '"<Car: None>" needs to have a value for field "car" '
+            'before this many-to-many relationship can be used.'
+        ):
+            c.drivers._remove_items('car', 'drivers', self.driver)
+
+    def test_clear_empty(self):
+        c = Car.objects.create(make=None)
+        c.save()
```
Comment (olibook) [inline tests/m2m_through_regress/tests.py]: save is not needed : objects.create returns an already saved instance. Tests succeed without the save

**Comment by akuzminov** [comment]
Is everything alright with this PR?

File: tests/m2m_through_regress/tests.py [hunk 7]
```diff
@@ -202,7 +203,11 @@ def test_add(self):
 
     def test_add_null(self):
         nullcar = Car.objects.create(make=None)
-        with self.assertRaises(ValueError):
+        with self.assertRaisesMessage(
+            ValueError,
+            '"<Car: None>" needs to have a value for field "car" '
+            'before this many-to-many relationship can be used.'
+        ):
             nullcar.drivers._add_items('car', 'driver', self.unused_driver)
```
Comment (Ian-Foote) [inline tests/m2m_through_regress/tests.py]: This error message is confusing. A `Car` instance doesn't have a `car` field. Should it refer to the `make` field instead? The `CarDriver` through table has a `car` foreign key with `to_field='make'`.

**Comment by akuzminov** [comment]
I have updated PR with a new error message

**Comment by timgraham** [comment]
I merged the field name fixes in 39a884380226e3ae372cdc93b7227ad2c8e7b086 and b8741c005873396de24a085fe9ab13d2348a306b, thanks.  Looking at the rest of it, I'm not sure what happened to much of Anssi's original patch (https://github.com/django/django/compare/master...akaariai:ticket_17541). It seems that only the many-to-many changes survived the rebase?

**Comment by timgraham** [comment]
Hi Andrey, do you plan to continue? Some others are asking on the ticket if they can work on this.

**Comment by timgraham** [comment]
Closing due to inactivity.

**Pull request closed without merging**